### PR TITLE
fix: split scale domains into explicit and implicit.

### LINKED ISSUE #3153: Cannot specify custom domain (with order) for a layered plot
I was trying to create a workaround for https://github.com/vega/vega-lite/issues/3152
and realize that we cannot specify custom domain (with order) for a layered plot yet. 

```
{
  "$schema": "https://vega.github.io/schema/vega-lite/v2.json",
  "data": {
    "values": [
      {
        "value": 1,
        "label": "blue"
      },
      {
        "value": 2,
        "label": "green"
      },
      {
        "value": 3,
        "label": "yellow"
      },
      {
        "value": 4,
        "label": "red"
      }
    ]
  },
  "layer": [
    {
      "encoding": {
        "x": {
          "field": "label",
          "scale": {
            "domain": [
              "blue",
              "green",
              "yellow",
              "red"
            ]
          },
          "type": "ordinal"
        },
        "y": {
          "field": "value",
          "scale": {
            "zero": false
          },
          "type": "quantitative"
        },
        "order": {
          "field": "value",
          "type": "quantitative"
        }
      },
      "mark": "line"
    },
    {
      "encoding": {
        "x": {
          "field": "label",
          "type": "ordinal"
        },
        "y": {
          "field": "value",
          "scale": {
            "zero": false
          },
          "type": "quantitative"
        }
      },
      "mark": "point"
    }
  ]
}
```

The fact that the output Vega x-scale has `sort: true` make it no longer respect the custom order

```
"domain": {
  "fields": [
     ["blue","green","yellow","red"],
     {"data": "data_2","field": "label"}
  ],
  "sort": true
}
```

### PR DESCRIPTION
**Pull request by domoritz**
fixes #3153

image

### CODE REVIEW

File: src/compile/scale/component.ts [hunk 1]
```diff
@@ -29,15 +27,11 @@ export class ScaleComponent extends Split<ScaleComponentProps> {
   /**
    * Whether the scale definitely includes zero in the domain
    */
-  public get domainDefinitelyIncludesZero() {
+  public domainDefinitelyIncludesZero() {
```
Comment (kanitw) [inline src/compile/scale/component.ts]: lol why do we have to make it no longer a getter?

File: src/compile/scale/component.ts [hunk 2]
```diff
@@ -9,15 +9,13 @@ import {Explicit, Split} from '../split';
  * All VgDomain property except domain.
  * (We exclude domain as we have a special "domains" array that allow us merge them all at once in assemble.)
  */
-// TODO: also exclude domainRaw and property implement the right scaleComponent for selection domain
+// TODO: also exclude domainRaw and properly implement the right scaleComponent for selection domain
 
-export type ScaleComponentProps = Omit<VgScale, 'domain'>;
+export type ScaleComponentProps = Omit<VgScale, 'domain'> & {domains: VgNonUnionDomain[]};
```
Comment (kanitw) [inline src/compile/scale/component.ts]: minor: technically, it's a component for scale domain. Given it's a domain of a single scale, then `domain` (singular) is more appropriate?

File: src/compile/split.ts [hunk 3]
```diff
@@ -96,13 +96,10 @@ export function makeImplicit<T>(value: T): Explicit<T> {
   };
 }
 
+export type PropertyOf = 'scale' | 'axis' | 'legend' | '';
```
Comment (kanitw) [inline src/compile/split.ts]: propertyOf as a variable makes sense, but a bit confusing as a type. 

How about `parentProperty: SplitParentProperty` or something similar?

File: src/compile/scale/component.ts (hunk 1 above)
Comment (domoritz) [reply src/compile/scale/component.ts]: We actually compute something pretty complicated here. Using a getter feels like an antipattern to me.

File: src/compile/scale/component.ts (hunk 2 above)
Comment (domoritz) [reply src/compile/scale/component.ts]: Before parse, this is not the domain yet but an array of domains that gets merged unioned into one vega domain. The types are also incompatible so it's better to have a different name, imho.

**Pull request merged**